[FLINK-31889] Add documentation for implementing/loading enrichers

pgaref opened the pull request:
[https://issues.apache.org/jira/browse/FLINK-31889](https://issues.apache.org/jira/browse/FLINK-31889)

FailureEnrichers documentation

dmvk commented on docs/content/docs/deployment/advanced/failure_enrichers.md:
```diff
@@ -0,0 +1,53 @@
+
+To implement a custom plugin for your use-case:
+
+- Add your own FailureEnricher by implementing the `org.apache.flink.core.failure.FailureEnricher` interface.
+
+- Add your own FailureEnricherFactory by implementing the `org.apache.flink.core.failure.FailureEnricherFactory` interface.
```
It would be neat to add links into javadocs.

dmvk commented on docs/content/docs/deployment/advanced/failure_enrichers.md:
```diff
@@ -0,0 +1,53 @@
+
+As a plugin will be loaded in different classloader, log4j is not able to initialized correctly in your plugin. In this case,
+you should add the config below in `flink-conf.yaml`:
+- `plugin.classloader.parent-first-patterns.additional: org.slf4j`
```
This feels weird, you shouldn't bundle slf4j in the plugin

dmvk replied:
Also `org.apache.flink.configuration.CoreOptions#PARENT_FIRST_LOGGING_PATTERNS` already contains it

wangzzu commented on docs/content/docs/deployment/advanced/failure_enrichers.md:
```diff
@@ -0,0 +1,93 @@
+
+- Add your own FailureEnricherFactory by implementing the {{< gh_link file="/flink-core/src/main/java/org/apache/flink/core/failure/FailureEnricherFactory.java" name="FailureEnricherFactory" >}} interface.
+
+- Add a service entry. Create a file `META-INF/services/org.apache.flink.core.failure.FailureEnricherFactory` which contains the class name of your failure enricher factory class (see [Java Service Loader](https://docs.oracle.com/javase/8/docs/api/java/util/ServiceLoader.html) docs for more details).
```
maybe use `https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/ServiceLoader.html` is better, now flink has spported jdk17

wangzzu replied:
Here I think the user needs to be reminded that the file name must be named `org.apache.flink.core.failure.FailureEnricherFactory`

wangzzu commented on docs/content/docs/deployment/advanced/failure_enrichers.md:
````diff
@@ -0,0 +1,93 @@
+  If this configuration is empty, NO enrichers will be started. Example:
+```
+    jobmanager.failure-enrichers = org.apache.flink.test.plugin.jar.failure.CustomEnricher
+```
````
Should we add a chapter here to explain how to confirm that this mechanism is in effect?

wangzzu reviewed (commented):
Overall LGTM, some little comments

pgaref commented:
> Overall LGTM, some little comments

Thanks for taking a look @wangzzu!
Let me know how you feel about the latest changes!

wangzzu reviewed (approved):
LGTM cc @dmvk

zentol commented on docs/content/docs/deployment/advanced/failure_enrichers.md:
```diff
@@ -0,0 +1,112 @@
+See [Flink Plugin]({% link deployment/filesystems/plugins.md %}) for more details.
+
+{{< hint warning >}}
+Note that every FailureEnricher should have defined a set of {{< gh_link file="/flink-core/src/main/java/org/apache/flink/core/failure/FailureEnricher.java" name="output keys" >}} that may be associated with values. This set of keys has to be unique across enrichers otherwise may be ignored.
```
> otherwise may be ignored

Otherwise _what_ may be ignored?

zentol replied:
I don't understand what the implications "ignoring values" is. In practice all that happens if that one enricher would override the label of another one, right? Can we just write that?

zentol replied:
oh wow all enrichers that clash in any way are just completely dropped if there is a clash. That doesn't seem to match the documented behavior.

pgaref replied:
When enrichers output keys overlap, only the first one is loaded and the rest are dropped with an Error message -- that was the meaning of `This set of keys has to be unique across enrichers otherwise may be ignored.` 
Rephrasing

zentol replied:
> When enrichers output keys overlap, only the first one is loaded and the rest are dropped with an Error message

That doesn't seem to match `org.apache.flink.runtime.failure.FailureEnricherUtils#filterInvalidEnrichers`. `invalidEnrichers` will contain all enrichers that clashed, and those are all thrown out.

```
        final Set<Class<?>> invalidEnrichers =
                enrichersByKey.entrySet().stream()
                        .filter(entry -> entry.getValue().size() > 1)
                        .flatMap(
                                entry -> {
                                    LOG.warn(
                                            "Following enrichers have have registered duplicate output key [%s] and will be ignored: {}.",
                                            entry.getValue().stream()
                                                    .map(Class::getName)
                                                    .collect(Collectors.joining(", ")));
                                    return entry.getValue().stream();
                                })
                        .collect(Collectors.toSet());
```

`org.apache.flink.runtime.failure.FailureEnricherUtilsTest#testValidatedEnrichersWithInvalidEntries` confirms that, as it uses 3 enrichers, with 2 having clashing keys, and asserting that only 1 is loaded.

pgaref replied:
You are right, that was the initial logic that we then reworked before merging.
Updating README accordingly.

zentol commented on docs/content/docs/deployment/advanced/failure_enrichers.md:
```diff
@@ -0,0 +1,112 @@
+
+### Configuration
+
+JobManager loads FailureEnricher plugins at startup. To make sure your FailureEnrichers are loaded:
```
```suggestion
The JobManager loads FailureEnricher plugins at startup. To make sure your FailureEnrichers are loaded:
```

zentol commented on docs/content/docs/deployment/advanced/failure_enrichers.md:
````diff
@@ -0,0 +1,112 @@
+* All class names should be defined as part of {{< javadoc file="org/apache/flink/configuration/JobManagerOptions.html#FAILURE_ENRICHERS_LIST" name="jobmanager.failure-enrichers">}} configuration.
+  If this configuration is empty, NO enrichers will be started. Example:
+```
+    jobmanager.failure-enrichers = org.apache.flink.test.plugin.jar.failure.CustomEnricher
````
This would be more complete if the CustomEnricher example included a `package`

zentol replied:
Did you miss this one?

pgaref replied:
`org.apache.flink.test.plugin.jar.failure` is the package? Not sure what you are referring to here

zentol replied:
Where did you add the package to the example implementation?

pgaref replied:
fixed

zentol commented on docs/content/docs/deployment/advanced/failure_enrichers.md:
````diff
@@ -0,0 +1,112 @@
+
+Moreover, you can query the JobManager's Rest API using the following endpoint:
+```
+    http://localhost:8081/jobs/<jobid>/exceptions
````
You could link to the REST API docs instead.

zentol commented on docs/content/docs/deployment/advanced/failure_enrichers.md:
```diff
@@ -0,0 +1,112 @@
+The goal is to enable developers implement their own failure enrichment plugins to categorize job failures, expose custom metrics, make calls to external notification systems, and more.
+
+FailureEnrichers are triggered every time an exception is reported at runtime by the JobManager.
+Every FailureEnricher may asynchronously return labels associated with the failure that are then exposed via the JobManager's Rest interface (e.g., a 'type:System' label implying the failure is categorized as a system error).
```
```suggestion
Every FailureEnricher may asynchronously return labels associated with the failure that are then exposed via the JobManager's REST API (e.g., a 'type:System' label implying the failure is categorized as a system error).
```

zentol commented on docs/content/docs/deployment/advanced/failure_enrichers.md:
```diff
@@ -0,0 +1,112 @@
+-->
+
+## Custom failure enrichers
+Flink provides a pluggable interface for users to register their custom logic and enrich failures with extra metadata labels (string kv pairs).
```
```suggestion
Flink provides a pluggable interface for users to register their custom logic and enrich failures with extra metadata labels (string key-value pairs).
```

zentol commented on docs/content/docs/deployment/advanced/failure_enrichers.md:
```diff
@@ -0,0 +1,112 @@
+
+## Custom failure enrichers
+Flink provides a pluggable interface for users to register their custom logic and enrich failures with extra metadata labels (string kv pairs).
+The goal is to enable developers implement their own failure enrichment plugins to categorize job failures, expose custom metrics, make calls to external notification systems, and more.
```
```suggestion
This enables users to implement their own failure enrichment plugins to categorize job failures, expose custom metrics, or make calls to external notification systems.
```

Consistent target audience, clarify that you _can_ do that stuff and that it's not just a "goal" (which _may_ not have been reached).
I'd also remove "and more" because it doesn't add any value. You don't want users guessing what this can be used for.

zentol commented on docs/content/docs/deployment/advanced/failure_enrichers.md:
```diff
@@ -0,0 +1,112 @@
+- Add a service entry. Create a file `META-INF/services/org.apache.flink.core.failure.FailureEnricherFactory` which contains the class name of your failure enricher factory class (see [Java Service Loader](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/ServiceLoader.html) docs for more details).
+
+
+Then, create a jar which includes your `FailureEnricher`, `FailureEnricherFactory`, `META-INF/services/` and all the external dependencies.
```
```suggestion
Then, create a jar which includes your `FailureEnricher`, `FailureEnricherFactory`, `META-INF/services/` and all external dependencies.
```

zentol commented on docs/content/docs/deployment/advanced/failure_enrichers.md:
```diff
@@ -0,0 +1,112 @@
+### Configuration
+
+JobManager loads FailureEnricher plugins at startup. To make sure your FailureEnrichers are loaded:
+* All class names should be defined as part of {{< javadoc file="org/apache/flink/configuration/JobManagerOptions.html#FAILURE_ENRICHERS_LIST" name="jobmanager.failure-enrichers">}} configuration.
```
Feels weird to have a list containing a single item.

zentol replied:
Please link to the config docs instead of the javadocs.

zentol reviewed (changes requested):
This needs a copy in the chinese version of the docs.

pgaref commented:
> This needs a copy in the chinese version of the docs. (that may be in english; translation is a follow-up)

Thanks for the review @zentol ! 
Addressed your comments, also created a copy as the chinese version of the docs that will be addressed as part of https://issues.apache.org/jira/browse/FLINK-33061 by @wangzzu

zentol commented on docs/content/docs/deployment/advanced/failure_enrichers.md:
````diff
@@ -0,0 +1,106 @@
+The JobManager loads FailureEnricher plugins at startup. To make sure your FailureEnrichers are loaded all class names should be defined as part of [jobmanager.failure-enrichers configuration]({{< ref "docs/deployment/config#jobmanager-failure-enrichers" >}}).
+  If this configuration is empty, NO enrichers will be started. Example:
+```
+    jobmanager.failure-enrichers = org.apache.flink.test.plugin.jar.failure.CustomEnricher
````
side-note: If multiple enrichers are used this option can become a bit cumbersome for users, and a structure like we use for metrics reporters could've solved that (and paved a way to make them configurable!)

```
jobmanager.failure-enrichers.enricher1.class = org.apache.flink.test.plugin.jar.failure.CustomEnricher
```

zentol commented on docs/content/docs/deployment/advanced/failure_enrichers.md:
````diff
@@ -0,0 +1,106 @@
+The JobManager loads FailureEnricher plugins at startup. To make sure your FailureEnrichers are loaded all class names should be defined as part of [jobmanager.failure-enrichers configuration]({{< ref "docs/deployment/config#jobmanager-failure-enrichers" >}}).
+  If this configuration is empty, NO enrichers will be started. Example:
+```
+    jobmanager.failure-enrichers = org.apache.flink.test.plugin.jar.failure.CustomEnricher
````
Is it correct that you configure the enricher itself, and not the factory? How does the factory lookup work? Do we just load all factories, have them all instantiate the enricher and do classname checks to find the right one...?

zentol replied:
Oh my it does work like that :(

Better hope the enrichers don't setup any resources in their constructor :/

pgaref replied:
So what you are implying here is by filtering on the Factory level we could avoid instantiating unwanted enrichers that could be potentialy setting up resources? 

Happy to add that as a followup ticket and follow an approach similar to MetricsReporter setup

pgaref commented on docs/content.zh/docs/deployment/advanced/failure_enrichers.md:
```diff
@@ -0,0 +1,106 @@
+
+## Custom failure enrichers
+Flink provides a pluggable interface for users to register their custom logic and enrich failures with extra metadata labels (string key-value pairs).
+This enables users to implement their own failure enrichment plugins to categorize job failures, expose custom metrics, or make calls to external notification systems.
```
like a setup/close methods? Main functionality is enriching/labeling exceptions but that would be a useful addition indeed. Lets see what users do with it first :)

pgaref commented:
@zentol just addressed latest comments -- please let me know what you think
